File: AprilTagScanner/Services/DetectionEngine.cs
using AprilTagScanner.Models;
using OpenCvSharp;
using OpenCvSharp.Aruco;

namespace AprilTagScanner.Services;

public sealed class DetectionEngine : IDisposable
{
    private readonly object _lock = new();
    private readonly Thread _worker;
    private volatile bool _running = true;
    private volatile bool _busy;

    private Mat? _pendingBgr;
    private int _detectWidth;
    private IReadOnlyList<string> _families = [];
    private bool _probe;
    private float _decimate = 1.5f;

    private List<DetectedTag> _lastTags = [];
    private double _lastMs;
    private int _cacheVersion;

    private readonly Dictionary<(string Family, bool Probe), Dictionary> _dictCache = [];

    public DetectionEngine()
    {
        _worker = new Thread(WorkerLoop) { IsBackground = true, Name = "AprilTagDetect" };
        _worker.Start();
    }

    public void BumpCache() => Interlocked.Increment(ref _cacheVersion);

    public void Configure(float decimate) => _decimate = decimate;

    public bool TrySubmit(Mat bgrFrame, int detectWidth, IReadOnlyList<string> families, bool probe)
    {
        if (families.Count == 0)
            return false;

        Mat clone;
        try
        {
            clone = bgrFrame.Clone();
        }
        catch
        {
            return false;
        }

        lock (_lock)
        {
            if (_busy)
            {
                clone.Dispose();
                return false;
            }

            _pendingBgr?.Dispose();
            _pendingBgr = clone;
            _detectWidth = detectWidth;
            _families = families;
            _probe = probe;
            _busy = true;
            Monitor.Pulse(_lock);
            return true;
        }
    }

    public (IReadOnlyList<DetectedTag> Tags, double Ms, bool Busy) Snapshot()
    {
        lock (_lock)
        {
            return (_lastTags.ToList(), _lastMs, _busy);
        }
    }

    private static Mat PrepareGray(Mat bgr, int detectWidth, out float invScale)
    {
        var w = bgr.Width;
        var h = bgr.Height;
        var scale = Math.Min(1.0, detectWidth / (double)w);
        if (scale < 1.0)
        {
            var small = new Mat();
            Cv2.Resize(
                bgr,
                small,
                new Size((int)(w * scale), (int)(h * scale)),
                interpolation: InterpolationFlags.Area
            );
            var gray = new Mat();
            Cv2.CvtColor(small, gray, ColorConversionCodes.BGR2GRAY);
            small.Dispose();
            invScale = (float)(1.0 / scale);
            return gray;
        }

        invScale = 1f;
        var direct = new Mat();
        Cv2.CvtColor(bgr, direct, ColorConversionCodes.BGR2GRAY);
        return direct;
    }

    private Dictionary GetDictionary(string family, bool probe)
    {
        var key = (family, probe);
        if (_dictCache.TryGetValue(key, out var dict))
            return dict;

        dict = CvAruco.GetPredefinedDictionary(TagFamilyCatalog.ToDictionary(family));
        _dictCache[key] = dict;
        return dict;
    }

    private void WorkerLoop()
    {
        var localCacheVersion = -1;
        var parameters = new DetectorParameters();

        while (_running)
        {
            Mat? bgr = null;
            int detectWidth;
            IReadOnlyList<string> families;
            bool probe;

            lock (_lock)
            {
                while (_running && _pendingBgr == null)
                    Monitor.Wait(_lock, 50);

                if (!_running)
                    break;

                bgr = _pendingBgr;
                _pendingBgr = null;
                detectWidth = _detectWidth;
                families = _families;
                probe = _probe;
            }

            if (bgr == null)
                continue;

            Mat? gray = null;
            var sw = System.Diagnostics.Stopwatch.StartNew();
            var found = new List<DetectedTag>();

            try
            {
                gray = PrepareGray(bgr, detectWidth, out var invScale);

                if (localCacheVersion != _cacheVersion)
                {
                    foreach (var d in _dictCache.Values)
                        d.Dispose();
                    _dictCache.Clear();
                    localCacheVersion = _cacheVersion;
                }

                foreach (var family in families)
                {
                    var dict = GetDictionary(family, probe);
                    CvAruco.DetectMarkers(gray, dict, out var corners, out var ids, parameters, out _);
                    if (ids == null || corners == null || ids.Length == 0)
                        continue;

                    for (var i = 0; i < ids.Length; i++)
                    {
                        var scaled = corners[i]
                            .Select(p => new Point2f(p.X * invScale, p.Y * invScale))
                            .ToArray();
                        found.Add(new DetectedTag
                        {
                            Family = family,
                            Id = ids[i],
                            Corners = scaled,
                        });
                    }
                }
            }
            catch
            {
                found.Clear();
            }
            finally
            {
                gray?.Dispose();
                bgr.Dispose();
            }

            sw.Stop();
            lock (_lock)
            {
                _lastTags = found;
                _lastMs = sw.Elapsed.TotalMilliseconds;
                _busy = false;
            }
        }
    }

    public void Dispose()
    {
        _running = false;
        lock (_lock)
        {
            _pendingBgr?.Dispose();
            _pendingBgr = null;
            Monitor.Pulse(_lock);
        }

        if (_worker.IsAlive)
            _worker.Join(TimeSpan.FromSeconds(1));

        foreach (var d in _dictCache.Values)
            d.Dispose();
        _dictCache.Clear();
    }
}

File: AprilTagScannerAndroid/app/src/main/kotlin/com/scanapriltag/services/DetectionEngine.kt
package com.scanapriltag.services

import com.scanapriltag.models.DetectedTag
import org.opencv.core.Mat
import org.opencv.core.Point
import org.opencv.core.Size
import org.opencv.imgproc.Imgproc
import org.opencv.objdetect.ArucoDetector
import org.opencv.objdetect.DetectorParameters
import org.opencv.objdetect.Dictionary
import org.opencv.objdetect.Objdetect
import java.util.concurrent.atomic.AtomicInteger

class DetectionEngine : AutoCloseable {
    private val lock = Any()
    private val worker = Thread(::workerLoop, "AprilTagDetect").apply {
        isDaemon = true
        start()
    }

    @Volatile
    private var running = true

    @Volatile
    private var busy = false

    private var pendingBgr: Mat? = null
    private var detectWidth = 960
    private var families: List<String> = emptyList()
    private var probe = false

    @Volatile
    private var decimate = 1.5f

    private var lastTags: List<DetectedTag> = emptyList()
    private var lastMs = 0.0

    private val cacheVersion = AtomicInteger(0)
    private val detectorCache = mutableMapOf<String, ArucoDetector>()

    fun bumpCache() {
        cacheVersion.incrementAndGet()
    }

    fun configure(decimateValue: Float) {
        if (decimate == decimateValue) return
        decimate = decimateValue
        bumpCache()
    }

    fun trySubmit(bgrFrame: Mat, detectWidthValue: Int, familiesValue: List<String>, probeMode: Boolean): Boolean {
        if (familiesValue.isEmpty()) return false

        val clone = try {
            bgrFrame.clone()
        } catch (_: Exception) {
            return false
        }

        synchronized(lock) {
            if (busy) {
                clone.release()
                return false
            }
            pendingBgr?.release()
            pendingBgr = clone
            detectWidth = detectWidthValue
            families = familiesValue
            probe = probeMode
            busy = true
            (lock as Object).notifyAll()
            return true
        }
    }

    fun snapshot(): Triple<List<DetectedTag>, Double, Boolean> =
        synchronized(lock) {
            Triple(lastTags.toList(), lastMs, busy)
        }

    private fun prepareGray(bgr: Mat, width: Int): Pair<Mat, Float> {
        val w = bgr.cols()
        val scale = minOf(1.0, width.toDouble() / w)
        return if (scale < 1.0) {
            val small = Mat()
            Imgproc.resize(
                bgr,
                small,
                Size(w * scale, bgr.rows() * scale),
                0.0,
                0.0,
                Imgproc.INTER_AREA,
            )
            val gray = Mat()
            Imgproc.cvtColor(small, gray, Imgproc.COLOR_BGR2GRAY)
            small.release()
            gray to (1.0f / scale.toFloat())
        } else {
            val gray = Mat()
            Imgproc.cvtColor(bgr, gray, Imgproc.COLOR_BGR2GRAY)
            gray to 1f
        }
    }

    private fun getDetector(family: String): ArucoDetector {
        val key = "$family@${decimate}"
        return detectorCache.getOrPut(key) {
            val dict: Dictionary = Objdetect.getPredefinedDictionary(TagFamilyCatalog.toDictionary(family))
            val params = DetectorParameters().apply {
                set_aprilTagQuadDecimate(decimate)
            }
            ArucoDetector(dict, params)
        }
    }

    private fun workerLoop() {
        var localCacheVersion = -1

        while (running) {
            val frame = dequeueFrame() ?: continue

            var gray: Mat? = null
            val start = System.nanoTime()
            val found = mutableListOf<DetectedTag>()

            try {
                val (grayMat, invScale) = prepareGray(frame.bgr, frame.width)
                gray = grayMat

                val version = cacheVersion.get()
                if (localCacheVersion != version) {
                    detectorCache.clear()
                    localCacheVersion = version
                }

                for (family in frame.families) {
                    val detector = getDetector(family)
                    val corners = mutableListOf<Mat>()
                    val ids = Mat()
                    val rejected = mutableListOf<Mat>()

                    detector.detectMarkers(gray, corners, ids, rejected)
                    rejected.forEach { it.release() }

                    if (ids.empty()) {
                        ids.release()
                        corners.forEach { it.release() }
                        continue
                    }

                    for (i in 0 until ids.rows()) {
                        val cornerMat = corners[i]
                        val scaled = Array(4) { idx ->
                            val p = cornerMat.get(0, idx)
                            Point(p[0] * invScale, p[1] * invScale)
                        }
                        found.add(
                            DetectedTag(
                                family = family,
                                id = ids.get(i, 0)[0].toInt(),
                                corners = scaled,
                            ),
                        )
                        cornerMat.release()
                    }
                    ids.release()
                }
            } catch (_: Exception) {
                found.clear()
            } finally {
                gray?.release()
                frame.bgr.release()
            }

            val elapsedMs = (System.nanoTime() - start) / 1_000_000.0
            synchronized(lock) {
                lastTags = found
                lastMs = elapsedMs
                busy = false
            }
        }
    }

    private data class PendingFrame(
        val bgr: Mat,
        val width: Int,
        val families: List<String>,
        val probe: Boolean,
    )

    private fun dequeueFrame(): PendingFrame? {
        synchronized(lock) {
            while (running && pendingBgr == null) {
                (lock as Object).wait(50)
            }
            if (!running) return null

            val bgr = pendingBgr ?: return null
            pendingBgr = null
            return PendingFrame(bgr, detectWidth, families.toList(), probe)
        }
    }

    override fun close() {
        running = false
        synchronized(lock) {
            pendingBgr?.release()
            pendingBgr = null
            (lock as Object).notifyAll()
        }
        worker.join(1000)
        detectorCache.clear()
    }
}
